secure/net/src/network_controller/outbound_handler.rs

```diff
@@ -29,84 +30,127 @@ impl OutboundHandler {
+            )
+            .await;
+            info!("Stopping outbound handler at {}", address.to_string());
+        });
+        Some(stop_signal_tx)
     }
 
-    fn process_one_outgoing_message(
-        outbound_handlers: Arc<Mutex<Vec<(Receiver<Message>, SocketAddr, MessageType)>>>,
-        network_clients: Arc<Mutex<HashMap<SocketAddr, NetworkClient>>>,
+    async fn process_one_outgoing_message(
+        outbound_handlers: Vec<(Receiver<Message>, SocketAddr, MessageType)>,
         socket_addr: &SocketAddr,
         inbound_handler: Arc<Mutex<InboundHandler>>,
+        grpc_clients: &mut HashMap<SocketAddr, GRPCNetworkMessageServiceClientWrapper>,
     ) {
-        let mut select = Select::new();
-        let handlers = outbound_handlers.lock().unwrap();
+        loop {
+            let mut select = Select::new();
+            for (receiver, _, _) in outbound_handlers.iter() {
+                select.recv(receiver);
+            }
 
-        for (receiver, _, _) in handlers.iter() {
-            select.recv(receiver);
-        }
-        let oper = select.select();
-        let index = oper.index();
-        let msg = oper.recv(&handlers[index].0).unwrap();
-        let remote_addr = &handlers[index].1;
-        let message_type = &handlers[index].2;
-        if remote_addr == socket_addr {
-            // If the remote address is the same as the local address, then we are sending a message to ourselves
-            // so we should just pass it to the inbound handler
-            inbound_handler
-                .lock()
-                .unwrap()
-                .send_incoming_message_to_handler(message_type, msg);
-            return;
-        }
-        let mut binding = network_clients.lock().unwrap();
-        let network_client = binding.get_mut(remote_addr).unwrap();
-        let msg = bcs::to_bytes(&NetworkMessage::new(
-            *socket_addr,
-            msg,
-            message_type.clone(),
-        ))
-        .unwrap();
+            let index;
+            let msg;
+            {
+                let oper = select.select();
+                index = oper.index();
+                match oper.recv(&outbound_handlers[index].0) {
+                    Ok(m) => {
+                        msg = m;
+                    },
+                    Err(e) => {
+                        warn!(
+                            "{:?} for outbound handler on {:?}. This can happen in shutdown,\
+                             but should not happen otherwise",
+                            e.to_string(),
+                            socket_addr
+                        );
+                        return;
+                    },
+                }
+            }
+
+            let remote_addr = &outbound_handlers[index].1;
+            let message_type = &outbound_handlers[index].2;
 
-        retry(fixed_retry_strategy(5, 20), || network_client.write(&msg)).unwrap();
+            if message_type.get_type() == "stop_task" {
```

Comment:
What is the use of this stop message?

Reply:
It is a clean way of exiting this task. NetworkController::shutdown calls this

Reply:
For 2 reasons:
(1) Runtime going out scope kills the task in a heavy handed way. That is, if the task in middle of something, it will die in between.
(2) Not an expert on tokio runtimes, but i have seen hangs sometime. That is, runtime object is not dropped until the task dies

Reply:
> (1) Runtime going out scope kills the task in a heavy handed way. That is, if the task in middle of something, it will die in between.


This shouldn't be a problem right as the run time going out of scope means we are shutting down the controller anyways.. 

Do you see the test hanging without the shutdown support?